Create merge comments action

dinhtungdu opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here -->

Fixes #7880

This PR:
- Creates a new action that acts as a commenter: it takes the output of various actions, then posts them to only one comment on the PR page.
- Updates existing action to set the comment content as an action output.
- Updates existing workflows to use the new action.

### Screenshots

<img width="911" alt="image" src="https://user-images.githubusercontent.com/5423135/208658934-d519de86-6b40-4587-99ab-c7e1a4e3103f.png">

<img width="916" alt="image" src="https://user-images.githubusercontent.com/5423135/208661815-3fc36309-6bd1-4e27-a970-46c4ab94b85c.png">


Or see https://github.com/woocommerce/woocommerce-blocks/pull/7973#issuecomment-1359252211

dinhtungdu committed:
create merge comments action

dinhtungdu committed:
add missing detector

dinhtungdu committed:
better detection and formatting

dinhtungdu committed:
fix default values for create-comments

dinhtungdu committed:
hidden separator

dinhtungdu committed:
add missing setOutput

dinhtungdu committed:
fix regex

dinhtungdu committed:
visible separator

dinhtungdu committed:
fix step id

dinhtungdu committed:
update compare scripts

dinhtungdu committed:
use markdown to prevent broken syntax after updating comment

dinhtungdu committed:
add a line break after the section  id to ensure heading works properly

dinhtungdu committed:
add line breaks around separator

dinhtungdu committed:
use heading 2 only

dinhtungdu committed:
update footer message

dinhtungdu committed:
smaller footer

dinhtungdu committed:
remove unnecessary id

dinhtungdu committed:
Merge branch 'trunk' into add/github-actions-combining-bot-comments-action

dinhtungdu committed:
more future proof separator

dinhtungdu committed:
add order input

dinhtungdu committed:
fancy name

dinhtungdu committed:
use hidden identifier to make update comment content easier

Aljullu commented on .github/comments-aggregator/action.yml:
```diff
@@ -0,0 +1,19 @@
+name: 'Merge Comments'
+description: 'Merge bot comment into one comment to keep PR page clean.'
```
Nit: should this be plural?
```suggestion
description: 'Merge bot comments into one comment to keep PR page clean.'
```

Aljullu commented on .github/comments-aggregator/action.yml:
```diff
@@ -0,0 +1,19 @@
+        description: 'Comment content.'
+        required: true
+    order:
+        description: 'Order of the comment'
```
Nit: I think we should either add a period at the end of each description or not. Currently it's inconsistent: `Comment content.` (has period) and `Order of the comment` (has no period).

Aljullu commented on .github/comments-aggregator/utils.js:
```diff
@@ -0,0 +1,89 @@
+const indentifier = `<!-- comments-aggregator -->`;
```
Typo: `identifier`.

Aljullu reviewed (approved):
I really liked the way you implemented this, @dinhtungdu! And very happy to get this merged to reduce the comments noise.

I left some minor comments, but pre-approving.

dinhtungdu committed:
address reviews by @Aljullu

dinhtungdu committed:
add readme file

dinhtungdu commented:
@Aljullu I addressed your review, thank you so much! I also added a README file to make using this action easier. Because of that, can you please take another look?

Aljullu commented on .github/comments-aggregator/README.md:
```diff
@@ -0,0 +1,57 @@
+
+## Usage
+
+This action is meant is be used as the poster/commenter. Instead of having
```
`This action is meant is be used` → `This action is meant to be used`

Aljullu commented on .github/comments-aggregator/README.md:
````diff
@@ -0,0 +1,57 @@
+action output, then feed that output to `comments-aggregator` to let this action
+manage those comments for you.
+
+```yml
+            - name: Compare Assets
+              uses: ./.github/compare-assets
+              id: compare-assets
+              with:
+                  repo-token: '${{ secrets.GITHUB_TOKEN }}'
+                  compare: assets-list/assets.json
+                  create-comment: false
+            - name: Append report
+              uses: ./.github/comments-aggregator
+              with:
+                  repo-token: '${{ secrets.GITHUB_TOKEN }}'
+                  section-id: compare-assets-with-trunk
+                  content: ${{steps.compare-assets.outputs.comment}}
+```
````
Nit: I would move this code block one indentation level up, to match the code block below.

Aljullu reviewed (approved):
Two small comments, but LGTM.

dinhtungdu committed:
address code review

dinhtungdu committed:
use 80 cols width make markdown look weird on small editor windows

dinhtungdu committed:
fix example identation

dinhtungdu committed:
update footer

fabiankaegy commented:
Hey @dinhtungdu 👋 

Do you think the workflow that adds any changes in dependencies could ever be made publically available? That would be super useful for anyone building blocks :)

dinhtungdu commented:
Heyy @fabiankaegy,

> Do you think the workflow that adds any changes in dependencies could ever be made publically available? That would be super useful for anyone building blocks :)

If the action can help others, I'm 💯 agree with you.

@mikejolley as you're the author of the `Compare Assets` action, what do you think about publishing it as a dedicated action?

cc @nerrad

nerrad commented:
Good candidate for https://github.com/woocommerce/automations?

fabiankaegy commented:
After looking at the implementation here a bit more I think the version here is a little to purpose build for this repository setup and it doesn't use the `.asset.php` files which would be the case for most simple plugins. 

But seeing this inspired me to pick up work on an older exploration I had done that tries to do the same thing. I published it here: https://github.com/fabiankaegy/monitor-wordpress-dependencies-action
